server/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java

```diff
@@ -164,6 +212,26 @@ public int numberOfShardsWithState(ShardRoutingState... states) {
     public List<ShardRouting> shardsWithState(String index, ShardRoutingState... states) {
         List<ShardRouting> shards = new ArrayList<>();
 
+        if (states.length == 1) {
+            if(states[0] == ShardRoutingState.INITIALIZING) {
+                for (ShardRouting shardEntry : initializingShards.values()) {
+                    if (!shardEntry.getIndexName().equals(index)) {
+                        continue;
+                    }
+                    shards.add(shardEntry);
+                }
+                return shards;
+            } else if (states[0] == ShardRoutingState.RELOCATING) {
+                for (ShardRouting shardEntry : relocatingShards.values()) {
+                    if (!shardEntry.getIndexName().equals(index)) {
```

Comment:
We avoid unary `!` operators since they are too easy to miss, preferring this instead:

```suggestion
                    if (shardEntry.getIndexName().equals(index) == false) {
```